File: web/src/entities/door.ts
import { ImmutableVec2 } from '../math.js';
import { Entity } from '../entity.js';
import { Resource } from '../resource.js';
import type { Animation } from '../media/animation.js';

export class Door extends Entity {
  private readonly id: number;
  private opening = false;
  private closing = false;
  private doorHeight = 40;
  private frameCounter = 0;
  private readonly doorAnim: Animation;

  constructor(id: number) {
    super();
    this.id = id;
    this.setSize(new ImmutableVec2(10, 40));
    this.doorAnim = Resource.getAnimation('data/images/door.bmp', 1);
  }

  getLayer(): number { return 0; }

  override onRespawn(): void {
    this.opening = false;
    this.closing = false;
    this.doorHeight = 40;
  }

  override onButtonDown(id: number): void {
    if (this.id !== id) return;
    this.frameCounter = 0;
    this.opening = true;
    this.closing = false;
  }

  override onButtonUp(id: number): void {
    if (this.id !== id) return;
    this.closing = true;
    this.opening = false;
    this.frameCounter = 0;
  }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const x = Math.round(this.getDrawPositionX() + offsetX - this.getSize().x / 2);
    const y = Math.round(this.getDrawPositionY() + offsetY - this.getSize().y / 2);

    // Draw partial door using drawScaled to crop
    const img = this.doorAnim.getGameImage();
    if (img) {
      const sub = img.subImage(0, 0, 10, this.doorHeight);
      sub.draw(ctx, x, y);
    }
  }

  override update(): void {
    this.frameCounter++;

    if (this.doorHeight < 4) this.doorHeight = 4;

    if (this.opening && this.doorHeight !== 4) {
      this.doorHeight -= 2;
    } else if (this.opening) {
      this.frameCounter = 0;
    }

    if (this.closing && this.frameCounter % 5 === 0 && this.doorHeight !== 40) {
      this.doorHeight++;
    }

    const tileX = Math.floor((this.getPosition().x - this.getSize().x / 2) / 10);
    const tileY = Math.floor((this.getPosition().y - this.getSize().y / 2) / 10);

    for (let i = 0; i < 4; i++) {
      this.myRoom.setCollidable(tileX, tileY + i, i <= Math.floor(this.doorHeight / 10));
    }
  }
}

File: wii/src/entities/door.cpp
#include "entities/door.h"
#include "resource.h"
#include "media/animation.h"
#include "room.h"
#include <cmath>

Door::Door(int id)
    : id_(id) {
    setSize(Vec2(10, 40));
    doorAnim_ = Resource::getAnimation("data/images/door.bmp", 1);
}

void Door::onRespawn() {
    opening_ = false;
    closing_ = false;
    doorHeight_ = 40;
}

void Door::onButtonDown(int id) {
    if (id_ != id) return;
    doorFrameCounter_ = 0;
    opening_ = true;
    closing_ = false;
}

void Door::onButtonUp(int id) {
    if (id_ != id) return;
    closing_ = true;
    opening_ = false;
    doorFrameCounter_ = 0;
}

void Door::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(getDrawPositionX() + offsetX - size_.x / 2.0f));
    int y = static_cast<int>(std::round(getDrawPositionY() + offsetY - size_.y / 2.0f));

    GameImage img = doorAnim_->getGameImage();
    GameImage sub = img.subImage(0, 0, 10, doorHeight_);
    sub.draw(renderer, x, y);
}

void Door::update() {
    doorFrameCounter_++;

    if (doorHeight_ < 4) doorHeight_ = 4;

    if (opening_ && doorHeight_ != 4) {
        doorHeight_ -= 2;
    } else if (opening_) {
        doorFrameCounter_ = 0;
    }

    if (closing_ && doorFrameCounter_ % 5 == 0 && doorHeight_ != 40) {
        doorHeight_++;
    }

    int tileX = static_cast<int>(std::floor((position_.x - size_.x / 2.0f) / 10.0f));
    int tileY = static_cast<int>(std::floor((position_.y - size_.y / 2.0f) / 10.0f));

    for (int i = 0; i < 4; i++) {
        room_->setCollidable(tileX, tileY + i, i <= doorHeight_ / 10);
    }
}
